(BOLT-1284) Add skeleton bolt project directory

donoghuc opened the pull request:
This commit adds a skeleton bolt project directory to enable new users to quickly get started with bolt.

donoghuc committed:
(BOLT-1284) Add skeleton bolt project directory

This commit adds a skeleton bolt project directory to enable new users to quickly get started with bolt.

donoghuc commented on bolt.yaml:
```diff
@@ -0,0 +1,20 @@
+#   service-url: "https://foo.example.com:8143"
+#   cacert: /path/to/ca.pem
+#   task-environment: production
+# ssh:
```
For most skeleton files I added some commented out examples to stub out an idea for making a working example project that actually uses all the example files. Is that worth investing the time in to? Or do we want to limit this to simply a directory structure?

nicklewis replied:
I think that if we're going to have example content in these files, we should include some comments at least 
 explaining what the sections are for.

lucywyman replied:
If the main purpose for this is to have something folks can clone and get started with at contributor summit I don't think content is necessary. People will be familiar with the file names and can refer to docs, and we'll have lots of Bolt knowledgeable folks on hand to help. I think just having empty files is fine, and we can add content if we want to reuse this for something else.

nicklewis commented on README.md:
```diff
@@ -1 +1,63 @@
+The `Puppetfile` is used to install modules from the [Puppet Forge](https://forge.puppet.com/) or from a Git repository. Modules specified in the Puppetfile will be downloaded and installed into the `modules` directory. [Puppetfile reference](https://puppet.com/docs/bolt/latest/installing_tasks_from_the_forge.html#task-8928)
+
+### site-modules
+The `site-modules` directory containts the project specific `Bolt` modules. In this example there is a single module called `project`.
```
```suggestion
The `site-modules` directory contains the project specific `Bolt` modules. In this example there is a single module called `project`.
```

nicklewis commented on bolt.yaml:
```diff
@@ -0,0 +1,20 @@
+# pcp:
+#   service-url: "https://foo.example.com:8143"
+#   cacert: /path/to/ca.pem
+#   task-environment: production
```
We should probably omit `task-environment`, since that's quite an advanced use case

nicklewis commented on bolt.yaml:
```diff
@@ -0,0 +1,20 @@
+#   host-key-check: false
+#   private-key: ~/.ssh/bolt_id
+#   user: foo 
+#   interpreters:
```
`interpreters` is a complex topic that we maybe shouldn't cover in this file

nicklewis commented on bolt.yaml:
```diff
@@ -0,0 +1,20 @@
+#   interpreters:
+#      rb: /home/foo/.rbenv/versions/2.5.1/bin/ruby
+# log:
+#   console:
```
Configuring the console log level is a strange use case given BOLT-901, so we might want to omit that as well

nicklewis commented on data/common.yaml:
```diff
@@ -0,0 +1,2 @@
+# ---
+# hiera_data: "special data"
```
This key name is kind of meta and makes it read like this is a config file for hiera. We should use something more generic, and probably include a comment explaining the structure of the file

nicklewis commented on site-modules/project/manifests/init.pp:
```diff
@@ -0,0 +1,3 @@
+# class project(){
```
```suggestion
# class project() {
```

nicklewis commented on site-modules/project/plans/init.pp:
```diff
@@ -0,0 +1,3 @@
+# plan project(){
```
```suggestion
# plan project() {
```

nicklewis commented on site-modules/project/tasks/init.json:
```diff
@@ -0,0 +1,22 @@
+// {
```
Could we start them with a real task here? Maybe a simple "echo" or something.

Some of the features demonstrated in this sample metadata are a bit advanced, like `implementations`. Also, comments aren't supported here, which means this file will likely cause an error if you tried to run the task.

nicklewis commented on Puppetfile:
```diff
@@ -0,0 +1,4 @@
+#forge 'http://forge.puppetlabs.com'
+#
+#mod 'profiles', local: true
```
I don't think you really need to use `local` in a Boltdir, since those should be in `site-modules` anyhow

lucywyman commented on README.md:
```diff
@@ -1 +1,63 @@
+
+This repository is an example skeleton Bolt Project Directory or `Boltdir`. A `Boltdir` contains all of the configuration, code, and data loaded by Bolt. Read more about the project directory structure at the following [link.](https://puppet.com/docs/bolt/latest/bolt_project_directories.html#project-directories)
+
+## Directory structure
```
I think this should just be a link to the docs page, that way there's only 1 source of truth.

donoghuc replied:
I think it is important to leave this as is. I link to the docs about choosing an appropriate directory structure above. In terms of a "source of truth" this shows the directory structure of this repository. So the only change would be if we added or removed a file from this project.

lucywyman reviewed (changes requested):
2 comments, but should be quick to address! Thanks for doin' this Cas - I do think it'll make the getting started stuff at CS much easier.

adreyer commented on Puppetfile:
```diff
@@ -0,0 +1,8 @@
+## This is an example puppetfile. 
+
+## The commented out section shows the URL of the Puppet forge to use.
```
We should not include the "forge" directive. It's more or less deprecated and wrong.

donoghuc committed:
(BOLT-1284) Provide "empty" files with comment descriptions

Instead of including commented out example code, only include a comment with a short description of the file and a link to the documentation.

donoghuc commented:
Based on conversation in slack, I updated to only include a short description of each file and a link to docs as a comment in each file.
